query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,674 @@
+                },
+            }) as crate::Result<ConnectedEngine>
+        }
+        .instrument(span)
+        .await?;
+
+        *inner = Inner::Connected(engine);
+        Ok(())
+    }
+
+    pub async fn query(
+        &self,
+        body_str: *const c_char,
+        trace_str: *const c_char,
+        tx_id_str: *const c_char,
+    ) -> Result<String, Error> {
+        let dispatcher = self.logger.dispatcher();
+
+        async move {
+            let inner = self.inner.read().await;
+            let engine = inner.as_engine()?;
+
+            let body = get_cstr_safe(body_str).expect("Prisma engine execute body is missing");
+            let tx_id = get_cstr_safe(tx_id_str);
+            let trace = get_cstr_safe(trace_str).expect("Trace is needed");
+
+            let query = RequestBody::try_from_str(&body, engine.engine_protocol()).map_err(Error::from)?;
+
+            let span = if tx_id.is_none() {
+                tracing::info_span!("prisma:engine", user_facing = true)
+            } else {
+                Span::none()
+            };
+
+            let trace_id = telemetry::helpers::set_parent_context_from_json_str(&span, &trace);
+
+            async move {
+                let handler = RequestHandler::new(engine.executor(), engine.query_schema(), engine.engine_protocol());
+                let response = handler.handle(query, tx_id.map(TxId::from), trace_id).await;
+
+                let serde_span = tracing::info_span!("prisma:engine:response_json_serialization", user_facing = true);
+                Ok(serde_span.in_scope(|| serde_json::to_string(&response))?)
+            }
+            .instrument(span)
+            .await
+        }
+        .with_subscriber(dispatcher)
+        .await
+    }
+
+    /// Disconnect and drop the core. Can be reconnected later with `#connect`.
+    pub async fn disconnect(&self, trace_str: *const c_char) -> Result<(), Error> {
+        let trace = get_cstr_safe(trace_str).expect("Trace is needed");
+        // let dispatcher = self.logger.dispatcher();
+
+        // async_panic_to_js_error(async {
+        let span = tracing::info_span!("prisma:engine:disconnect");
+        let _ = telemetry::helpers::set_parent_context_from_json_str(&span, &trace);
+
+        // TODO: when using Node Drivers, we need to call Driver::close() here.
+
+        // async {
+        let mut inner = self.inner.write().await;
+        let engine = inner.as_engine()?;
+
+        let builder = EngineBuilder {
+            schema: engine.schema.clone(),
+            engine_protocol: engine.engine_protocol(),
+            native: EngineBuilderNative {
+                config_dir: engine.native.config_dir.clone(),
+                env: engine.native.env.clone(),
+            },
+        };
+
+        *inner = Inner::Builder(builder);
+
+        Ok(())
+        // }
+        // .instrument(span)
+        // .await
+        // })
+        // .with_subscriber(dispatcher)
+        // .await
```

Comment:
I was just struggling to get the code to work, this fell of the wayside. Let me try to add it back.

Reply:
Actually none of the calls have this span and dispatchers applied because I couldn't figure out how to get this working.

![Oscar Franco Screen 001217](https://github.com/prisma/prisma-engines/assets/1634213/885c1091-aec5-4760-92bf-757aadc38fd0)

What are they actually supposed to do? The dispatcher is for the logger, but the span? and how do I make them work without calling the async_panic_to_js_error present on the node version?

Reply:
I had the extra braces in there but the errors do not match
![Oscar Franco Screen 001218](https://github.com/prisma/prisma-engines/assets/1634213/a9f5a442-95df-413a-b8e8-02eb259df5b5)
![Oscar Franco Screen 001219](https://github.com/prisma/prisma-engines/assets/1634213/91dab287-41eb-4d57-86fb-805ba801e5e3)